mockkubeapiserver/hooks/namespace.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/mockkubeapiserver/storage"
+)
+
+type NamespaceHook struct {
+}
+
+func (s *NamespaceHook) OnWatchEvent(ev *storage.WatchEvent) {
+	switch ev.GroupKind() {
+	case schema.GroupKind{Kind: "Namespace"}:
+		s.namespaceChanged(ev)
+	}
+}
+
+// Fake the required transitions for Namespace objects so they become ready
+func (s *NamespaceHook) namespaceChanged(ev *storage.WatchEvent) {
+	u := ev.Unstructured()
+
+	// These changes seem to be done synchronously (similar to a mutating webhook)
+	labels := u.GetLabels()
+	name := u.GetName()
+	if labels["kubernetes.io/metadata.name"] != name {
```

Comment:
Yeah, I think it is legacy, but ... it's there.  I think we prefer field selectors now: https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/

Only a handful of fieldSelectors are implemented for some objects (though I think it would be cool if we supported arbitrary CEL expressions).  fieldSelector for `metadata.name` is implemented everywhere I believe, because that is how we can watch one object (you can only pass the `watch` parameter to LIST requests, IIRC)